### PR TITLE
fix(s3-deployment): `BucketDeployment` fails when bootstrap stack's `StagingBucket` is encrypted with customer managed KMS key

### LINKED ISSUE #25100: (aws-s3-deployment): bucket deployment fails when toolkit stack uses Customer KMS
### Describe the bug

When initialising the toolkit stack with `--bootstrap-kms-key-id` the bucket deployment fails.

### Expected Behavior

Bucket deployment copies the ressources to the desired destination

### Current Behavior

 The deployment fails and the logs read:

```
download failed: s3://cdk-<qualifier>-assets-<account>-<region>/<asset-hash>.zip to <tmp-folder> An error occurred (AccessDenied) when calling the GetObject operation: The ciphertext refers to a customer master key that does not exist, does not exist in this region, or you are not allowed to access.
```

### Reproduction Steps

Bootstrap with KMS: 

```shell
npx cdk bootstrap aws://$AWS_ACCOUNT_ID/$AWS_REGION --bootstrap-kms-key-id $KMS_KEY_ID
```

create a stack with a bucket deployment
```
import * as cdk from 'aws-cdk-lib';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as s3deploy from 'aws-cdk-lib/aws-s3-deployment';

export class MyStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Create an S3 bucket
    const bucket = new s3.Bucket(this, 'MyBucket', {
      bucketName: 'my-bucket-name'
    });

    // Deploy a text file to the bucket using BucketDeployment
    new s3deploy.BucketDeployment(this, 'DeployTextFile', {
      sources: [s3deploy.Source.asset('./path/to/text/file.txt')],
      destinationBucket: bucket,
      destinationKeyPrefix: 'path/to/deployed/text/file.txt'
    });
  }
}
```

### Possible Solution

The `handlerRole` of the BucketDeployment construct needs to have something like this:

```
    Key.fromKeyArn(this, 'cdk-toolkit-encryption-key', Stack.of(this).formatArn({
      region: Aws.REGION,
      service: 'kms',
      account: Aws.ACCOUNT_ID,
      resource: 'key',
      resourceName: Fn.importValue(`CdkBootstrap-${this.node.tryGetContext('@aws-cdk/core:bootstrapQualifier')}-FileAssetKeyArn`),
    }))
      // @ts-ignore
      .grantEncryptDecrypt(bucketDeployment.handlerRole);
```

### Additional Information/Context

Bootstrap Stack is updated to the most recent version

### CDK CLI Version

2.73.0

### Framework Version

_No response_

### Node.js Version

16.20.0

### OS

macos 13.2.1

### Language

Typescript

### Language Version

4.9.5

### Other information

_No response_

### PR DESCRIPTION
**Pull request by ghost**
### Issue #25100

Closes #25100.

### Reason for this change

When the CDK bootstrap stack's `StagingBucket` is encrypted with a customer managed KMS key whose key policy does not include wildcard KMS permissions similar to those of the S3 managed KMS key, `BucketDeployment` fails because the Lambda's execution role doesn't get the `kms:Decrypt` permission necessary to download from the bucket.

In addition, if one of the sources for `BucketDeployment` comes from the `Source.bucket` static method, and the bucket is an "out-of-app imported reference" created from `s3.Bucket.fromBucketName`, the bucket's `encryptionKey` attribute is `null` and the current code won't add the `kms:Decrypt` permission on the bucket's encryption key to the Lambda's execution role. If this bucket is additionally encrypted with a customer managed KMS key without sufficient resource-based policy, the deployment fails as well.

### Description of changes

It's not easy to make the code "just work" in every situation, because there's no way to pinpoint the source bucket's encryption key ARN without using another custom resource, which is a heavy-lifting and it's hard to give this new Lambda a reasonable and minimal set of execution role permissions.

Therefore, this PR resolves the issue by changing `BucketDeployment.handlerRole` from `private readonly` to `public readonly`, and adding documentations on how to handle errors resulting from "not authorized to perform: kms:Decrypt". The current code allows customizing `handlerRole` by passing in `BucketDeploymentProps.role`. This change makes the customization easier because users don't need to manually add the S3 permissions.

The only code change is on the visibility of `BucketDeployment.handlerRole`. All other changes are documentations.

I proposed 4 possible changes in my comment to Issue #25100, and only the first one (changing visibility) is pursued in this PR. The second one was abandoned because the CFN export `CdkBootstrap-hnb659fds-FileAssetKeyArn` is deprecated.

### Description of how you validated changes

I wrote a CDK app which uses the `BucketDeployment` construct. After manually adding relevant KMS permissions to the Lambda execution role, I verified that the bucket deployment worked in the following two scenarios:

- My personal account; bootstrap stack's `StagingBucket` encrypted with a custom KMS key which only has the default key policy.
- GoDaddy corporate account; `StagingBucket` encrypted with a KMS key from an AWS Organization management account.

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by kxue-godaddy** [comment]
Exemption Request

I'm requesting an exemption from updating test and snapshot files in this PR.

The only code change in this PR is changing `BucketDeployment.handlerRole` from `private` to `public`. Other parts of the PR are pure documentation explaining when and how to use this newly exposed attribute to add onto the Lambda's execution role permissions.

**Comment by GavinZZ** [comment]
Agree that no test is needed. Just have one question on the `ts-fixture` file change.

File: packages/aws-cdk-lib/rosetta/aws_s3_deployment/default.ts-fixture
Comment (kxue-godaddy) [reply packages/aws-cdk-lib/rosetta/aws_s3_deployment/default.ts-fixture:8]: Thanks for reviewing! @GavinZZ 

I'm following the guidelines on writing docs in the [Rosetta](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md#rosetta) section. Without adding these lines, the new code snippets I added won't compile (e.g. Line 98 of [packages/aws-cdk-lib/aws-s3-deployment/README.md](https://github.com/aws/aws-cdk/pull/29540/files#diff-15a309d25a720b5fa37e0acf9fa11e0638de4302d6b9cc578e0efcb54110aeafR98), which uses `iam.PolicyStatement`). The section also recommends "Utilize the default.ts-fixture that already exists rather than writing new .ts-fixture files", so I'm updating the default ts-fixture file.

It seems that the CI (when I was making the PR) doesn't check if Rosetta compiles, but since this PR is mainly doc changes, I want to make sure the updates can be reflected on the CDK doc site.

Comment (GavinZZ) [reply packages/aws-cdk-lib/rosetta/aws_s3_deployment/default.ts-fixture:8]: @kxue-godaddy thanks for getting back to me so quickly. I agree that the import statement for `iam` is necessary since it's used in the README file, but I don't think I see any usage of `kms` (other than the IAM actions). Would you please try removing the `kms` import statement and re-ran the command?

Comment (kxue-godaddy) [reply packages/aws-cdk-lib/rosetta/aws_s3_deployment/default.ts-fixture:8]: @GavinZZ 

I removed the `kms` line and reran the command. It shows the following error:

```
aws-cdk-lib.aws_s3_deployment.Source#bucket-example.ts:24:18 - error TS2304: Cannot find name 'kms'.

24   encryptionKey: kms.Key.fromKeyArn(
```

The `kms` import line is needed by [packages/aws-cdk-lib/aws-s3-deployment/lib/source.ts](https://github.com/aws/aws-cdk/pull/29540/files#diff-dc0c9fe2615ec0f6de32500a9fbe2d5e5ba20743dded75bfc835707fd180651cR85), where I added an example for `Source.bucket` via the `@example` tag.

Comment (GavinZZ) [reply packages/aws-cdk-lib/rosetta/aws_s3_deployment/default.ts-fixture:8]: oh nice, thank you. Didn't notice that there's a `kms` usage here.

**Review by GavinZZ — DISMISSED** [review]
LGTM, thank you for contributing and answering my questions.

**Comment by kxue-godaddy** [comment]
Looks like "Queue: Embarked in merge queue" was not successful because I didn't allow workflow runs in my fork repo. Just updated that I will trigger mergify again.

**Comment by kxue-godaddy** [comment]
@mergifyio refresh

**Comment by kxue-godaddy** [comment]
@GavinZZ 

Mergify reports an error for "Queue: Embarked in merge queue" after PR approval. I tried clicking the "Update branch" button and mergify dismissed your approval as stale. Could you approve again? The changes are the same. Sorry!

**Review by paulhcsun — APPROVED** [review]

**Pull request merged**
